Fer va estimation (#264)
opendr-eu/opendr: docs/reference/image_based_facial_emotion_estimation.md

#### `FacialEmotionLearner` constructor
```python
FacialEmotionLearner(self, lr, batch_size, temp_path, device, device_ind, validation_interval,
                     max_training_epoch, momentum, ensemble_size, base_path_experiment, name_experiment, dimensional_finetune, categorical_train,
                     base_path_to_dataset, max_tuning_epoch, diversify)
```

Constructor parameters:

- **lr**: *float, default=0.1*\
  Specifies the initial learning rate to be used during training.
- **batch_size**: *int, default=32*\
  Specifies number of samples to be bundled up in a batch during training.
  This heavily affects memory usage, adjust according to your system.
- **temp_path**: *str, default='temp'*\
  Specifies a path where the algorithm saves the checkpoints and onnx optimized model (if needed).
- **device**: *{'cpu', 'cuda'}, default='cuda'*\
  Specifies the device to be used.
- **device_ind**: *list, default=[0]*\
  List of GPU indices to be used if the device is 'cuda'.
- **validation_interval**: *int, default=1*\
  Specifies the validation interval.

> - **max_training_epoch**: *int, default=100*\

Reviewer: ```suggestion
- **max_training_epoch**: *int, default=2*\
```

Reviewer: according to learner

  Specifies the maximum number of epochs the training should run for.
- **momentum**: *float, default=0.9*\
  Specifies the momentum value used for optimizer.
- **ensemble_size**: *int, default=9*\
  Specifies the number of ensemble branches in the model.
- **base_path_experiment**: *str, default='./experiments/'*\
  Specifies the path in which the experimental results will be saved.
- **name_experiment**: *str, default='esr_9'*\
  String name for saving checkpoints.
- **dimensional_finetune**: *bool, default=True*\

>   Specifies if the model should be fintuned on dimensional data or not.

Reviewer: ```suggestion
  Specifies if the model should be fine-tuned on dimensional data or not.
```

> - **categorical_train**: *bool, default=True*\

Reviewer: ```suggestion
- **categorical_train**: *bool, default=False*\
```

  Specifies if the model should be trained on categorical data or not.
- **base_path_to_dataset**: *str, default=''./data/AffectNet''*\
  Specifies the dataset path.

> - **max_tuning_epoch**: *int, default=20*\

Reviewer: ```suggestion
- **max_tuning_epoch**: *int, default=1*\
```

  Specifies the maximum number of epochs the model should be finetuned on dimensional data.

> - **max_tuning_epoch**: *bool, default=False*\

Reviewer: ```suggestion
- **diversity**: *bool, default=False*\
```

  Specifies if the learner diversifies the features of different branches or not.

#### `FacialEmotionLearner.fit`
```python
FacialEmotionLearner.fit(self)
```

This method is used for training the algorithm on a train dataset and validating on a val dataset.

#### `FacialEmotionLearner.eval`
```python
FacialEmotionLearner.eval(self, eval_type, current_branch_on_training)
```

This method is used to evaluate a trained model on an evaluation dataset.
Returns a dictionary containing stats regarding evaluation.

Parameters:

- **eval_type**: *str, default='categorical'*\
  Specifies the type of data that model is evaluated on.
  It can be either categorical or dimensional data.
- **current_branch_on_training**: *int, default=0*\
  Specifies the index of trained branch which should be evaluated on validation data.

#### `FacialEmotionLearner.init_model`
```python
FacialEmotionLearner.init_model(self, num_branches)
```

This method is used to initialize the model.

Parameters:

> - **num_branches**: *int, default=9*\

Reviewer: ```suggestion
- **num_branches**: *int*\
```

Reviewer: no default value on the learner side

  Specifies the number of ensemble branches in the model. ESR_9 model is built by 9 branches by default.

#### `FacialEmotionLearner.infer`
```python
FacialEmotionLearner.infer(self, input_batch)
```

This method is used to perform inference on an image or a batch of images.
It returns dimensional emotion results and also the categorical emotion results as an object of `engine.target.Category` if a proper input object `engine.data.Image` is given.

Parameters:

- **input_batch**: *object***
  Object of type `engine.data.Image`. It also can be a list of Image objects, or a Torch tensor which will be converted to Image object.

#### `FacialEmotionLearner.save`
```python

> FacialEmotionLearner.save(self, state_dicts, base_path_to_save_model, current_branch_save)

Reviewer: ```suggestion
FacialEmotionLearner.save(self, state_dicts, base_path_to_save_model)
```

Reviewer: no such parameter

```
This method is used to save a trained model.

> Provided with the path (absolute or relative), it creates the "path" directory, if it does not already
> exist.

Reviewer: ```suggestion
Provided with the path (absolute or relative), it creates the "path" directory, if it does not already exist.
```

Inside this folder, the model is saved as "model_name.pt" and the metadata file as "model_name.json". If the directory already exists, the "model_name.pt" and "model_name.json" files are overwritten.

If [`self.optimize`](#FacialEmotionLearner.optimize) was run previously, it saves the optimized ONNX model in a similar fashion with an ".onnx" extension, by copying it from the self.temp_path it was saved previously during conversion.

Parameters:

- **state_dicts**: *object*\
  Object of type Python dictionary containing the trained model weights.
- **base_path_to_save_model**: *str*\
  Specifies the path in which the model will be saved.

> - **current_branch_save**: *int*\
>   Specifies the index of the model branch which should be saved.

Reviewer: ```suggestion
```

#### `FacialEmotionLearner.load`
```python
FacialEmotionLearner.load(self, ensemble_size, path_to_saved_network, file_name_base_network,
                        file_name_conv_branch, fix_backbone)
```

Loads the model from inside the directory of the path provided, using the metadata .json file included.

Parameters:

- **ensemble_size**: *int, default=9*\
  Specifies the number of ensemble branches in the model for which the pretrained weights should be loaded.

> - **path_to_saved_network**: *str*\

Reviewer: ```suggestion
- **path_to_saved_network**: *str, default="./trained_models/esr_9"*\
```

  Path of the model to be loaded.

> - **file_name_base_network**: *str*\

Reviewer: ```suggestion
- **file_name_base_network**: *str, default="Net-Base-Shared_Representations.pt"*\
```

  The file name of the base network to be loaded.

> - **file_name_conv_branch**: *str*\

Reviewer: ```suggestion
- **file_name_conv_branch**: *str, default="Net-Branch_{}.pt"*\
```

  The file name of the ensemble branch network to be loaded.

> - **fix_backbone**: *bool, default=True*\

Reviewer: ```suggestion
- **fix_backbone**: *bool*\
```
